src/main/scala/firrtl/transforms/DeadCodeElimination.scala

```diff
@@ -0,0 +1,244 @@
+        getDeps(value).foreach(ref => depGraph.addEdge(node, ref))
+      case DefWire(_, name, _) =>
+        depGraph.addVertex(LogicNode(mod.name, name))
+      case mem: DefMemory =>
+        // Treat DefMems as a node with outputs depending on the node and node depending on inputs
+				// From perpsective of the module or instance, MALE expressions are inputs, FEMALE are outputs
+        val memRef = WRef(mem.name, MemPortUtils.memType(mem), ExpKind, FEMALE)
+				val exprs = Utils.create_exps(memRef).groupBy(Utils.gender(_))
+				val sources = exprs.getOrElse(MALE, List.empty).flatMap(getDeps(_))
+				val sinks = exprs.getOrElse(FEMALE, List.empty).flatMap(getDeps(_))
+        val memNode = getDeps(memRef) match { case Seq(node) => node }
+        depGraph.addVertex(memNode)
+        sinks.foreach(sink => depGraph.addEdge(sink, memNode))
+        sources.foreach(source => depGraph.addEdge(memNode, source))
+      case Attach(_, exprs) => // Add edge between each expression
+        exprs.flatMap(getDeps(_)).toSet.subsets(2).map(_.toList).foreach {
+          case Seq(a, b) =>
+            depGraph.addEdge(a, b)
+            depGraph.addEdge(b, a)
+        }
+      case Connect(_, loc, expr) =>
+        // This match enforces the low Firrtl requirement of expanded connections
+        val node = getDeps(loc) match { case Seq(elt) => elt }
+        getDeps(expr).foreach(ref => depGraph.addEdge(node, ref))
+      // Simulation constructs are treated as top-level outputs
+      case Stop(_,_, clk, en) =>
+        Seq(clk, en).flatMap(getDeps(_)).foreach(ref => depGraph.addEdge(circuitSink, ref))
+      case Print(_, _, args, clk, en) =>
+        (args :+ clk :+ en).flatMap(getDeps(_)).foreach(ref => depGraph.addEdge(circuitSink, ref))
+      case Block(stmts) => stmts.foreach(onStmt(_))
+      case ignore @ (_: IsInvalid | _: WDefInstance | EmptyStmt) => // do nothing
+      case other => throw new Exception(s"Unexpected Statement $other")
+    }
+
+    // Add all ports as vertices
+    mod.ports.foreach { case Port(_, name, _, _: GroundType) =>
+      depGraph.addVertex(LogicNode(mod.name, name))
+    }
+    onStmt(mod.body)
+  }
+
+  // TODO Make immutable?
+  private def createDependencyGraph(instMaps: collection.Map[String, collection.Map[String, String]],
+                                    c: Circuit): MutableDiGraph[LogicNode] = {
+    val depGraph = new MutableDiGraph[LogicNode]
+    c.modules.foreach {
+      case mod: Module => setupDepGraph(depGraph, instMaps(mod.name))(mod)
+      case ext: ExtModule =>
+        // Connect all inputs to all outputs
+        val node = LogicNode(ext.name, ext.name)
+        val ports = ext.ports.groupBy(_.direction)
+        depGraph.addVertex(node)
+        ports.get(Output).foreach(_.foreach(output => depGraph.addEdge(LogicNode(ext.name, output.name), node)))
+        ports.get(Input).foreach(_.foreach(input => depGraph.addEdge(node, LogicNode(ext.name, input.name))))
```

Comment:
Some ExtModules should be preserved even if their outputs are unused (e.g., monitors) and some should be deleted (e.g., RAMs).  So, if don't-prune is the default, we'd need to add a DoTouch annotation.

Superficially, it seems better to prune by default and use the existing DontTouch annotation to treat them as sinks.